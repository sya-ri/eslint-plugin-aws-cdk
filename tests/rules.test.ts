import plugin from "../src/index.ts";
import * as path from "node:path";
import * as fs from "node:fs";

const ruleNames = Object.keys(plugin.rules);
const numberOfRules = 1;

describe("rules", () => {
  it("should have a corresponding doc for each rule", () => {
    ruleNames.forEach((rule) => {
      const docPath = path.resolve(__dirname, "../docs/rules", `${rule}.md`);

      if (!fs.existsSync(docPath)) {
        throw new Error(
          `Could not find documentation file for rule "${rule}" in path "${docPath}"`,
        );
      }
    });
  });

  it("should have a corresponding test for each rule", () => {
    ruleNames.forEach((rule) => {
      const testPath = path.resolve(__dirname, "./rules/", `${rule}.test.ts`);

      if (!fs.existsSync(testPath)) {
        throw new Error(
          `Could not find test file for rule "${rule}" in path "${testPath}"`,
        );
      }
    });
  });

  it("should have the correct amount of rules", () => {
    const { length } = ruleNames;

    if (length !== numberOfRules) {
      throw new Error(
        `There should be exactly ${numberOfRules} rules, but there are ${length}. If you've added a new rule, please update this number.`,
      );
    }
  });

  it("should export configs that refer to actual rules", () => {
    const expectAwsCdkPlugin = expect.objectContaining({
      meta: {
        name: "eslint-plugin-aws-cdk",
        version: expect.any(String),
      },
    });

    const recommendedConfigs = plugin.configs;

    expect(recommendedConfigs).toMatchSnapshot({
      "flat/recommended": { plugins: { "aws-cdk": expectAwsCdkPlugin } },
      "flat/all": { plugins: { "aws-cdk": expectAwsCdkPlugin } },
    });
    expect(Object.keys(recommendedConfigs)).toEqual([
      "all",
      "recommended",
      "flat/all",
      "flat/recommended",
    ]);
    expect(Object.keys(recommendedConfigs.all.rules)).toHaveLength(
      ruleNames.length,
    );
    const allConfigRules = Object.values(recommendedConfigs).flatMap((config) =>
      Object.keys(config.rules ?? {}),
    );

    allConfigRules.forEach((rule) => {
      const ruleNamePrefix = "aws-cdk/";
      const ruleName = rule.slice(ruleNamePrefix.length);

      expect(rule.startsWith(ruleNamePrefix)).toBe(true);
      expect(ruleNames).toContain(ruleName);
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      expect(() => require(`../src/rules/${ruleName}`)).not.toThrow();
    });
  });
});
