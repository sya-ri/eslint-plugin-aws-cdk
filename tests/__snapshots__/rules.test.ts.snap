// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rules should export configs that refer to actual rules 1`] = `
{
  "all": {
    "plugins": [
      "aws-cdk",
    ],
    "rules": {
      "aws-cdk/no-use-parent-construct": "error",
    },
  },
  "flat/all": {
    "plugins": {
      "aws-cdk": ObjectContaining {
        "meta": {
          "name": "eslint-plugin-aws-cdk",
          "version": Any<String>,
        },
      },
    },
    "rules": {
      "aws-cdk/no-use-parent-construct": "error",
    },
  },
  "flat/recommended": {
    "plugins": {
      "aws-cdk": ObjectContaining {
        "meta": {
          "name": "eslint-plugin-aws-cdk",
          "version": Any<String>,
        },
      },
    },
    "rules": {
      "aws-cdk/no-use-parent-construct": "error",
    },
  },
  "recommended": {
    "plugins": [
      "aws-cdk",
    ],
    "rules": {
      "aws-cdk/no-use-parent-construct": "error",
    },
  },
}
`;
